pykafka/managedbalancedconsumer.py

```diff
@@ -0,0 +1,368 @@
+
+    def stop(self):
+        """Stop this consumer
+
+        Should be called as part of a graceful shutdown
+        """
+        self._running = False
+        if self._consumer is not None:
+            self._consumer.stop()
+        self._group_coordinator.leave_managed_consumer_group(self._consumer_group,
+                                                             self._consumer_id)
+
+    def _send_heartbeat(self):
+        """Send a heartbeat request to the group coordinator and react to the response"""
+        res = self._group_coordinator.group_heartbeat(
+            self._consumer_group, self._generation_id, self._consumer_id)
+        if res.error_code == 0:
+            return
+        log.info("Error code %d encountered on heartbeat." % res.error_code)
+        if res.error_code in (IllegalGeneration.ERROR_CODE,
+                              RebalanceInProgress.ERROR_CODE,
+                              UnknownMemberId.ERROR_CODE):
+            pass
+        elif res.error_code in (GroupCoordinatorNotAvailable.ERROR_CODE,
+                                NotCoordinatorForGroup.ERROR_CODE):
+            self._group_coordinator = self._cluster.get_group_coordinator(
+                self._consumer_group)
+        elif res.error_code == GroupAuthorizationFailed.ERROR_CODE:
+            raise GroupAuthorizationFailed()
+        self._rebalance()
+
+    def _decide_partitions(self, participants, consumer_id=None):
+        """Decide which partitions belong to this consumer
+
+        Thin formatting wrapper around
+        `pykafka.balancedconsumer.BalancedConsumer._decide_partitions`
+        """
+        parts = super(ManagedBalancedConsumer, self)._decide_partitions(
+            participants, consumer_id=consumer_id)
+        return [p.id for p in parts]
+
+    def _update_member_assignment(self):
+        """Join a managed consumer group and start consuming assigned partitions
+
+        Equivalent to
+        `pykafka.balancedconsumer.BalancedConsumer._update_member_assignment`,
+        but uses the Kafka 0.9 Group Membership API instead of ZooKeeper to manage
+        group state
+        """
+        for i in range(self._rebalance_max_retries):
+            try:
+                members = self._join_group()
+                # generate partition assignments for each group member
+                # if this is not the leader, join_result.members will be empty
+                group_assignments = [
+                    MemberAssignment([
+                        (self._topic.name,
+                            self._decide_partitions(iterkeys(members),
+                                                    consumer_id=member_id))
+                    ], member_id=member_id) for member_id in members]
+
+                assignment = self._sync_group(group_assignments)
+                my_partitions = [p for p in itervalues(self._topic.partitions)
+                                 if p.id in assignment[0][1]]
```

Comment:
Just a nitpick: I think this could be written `[self._topic.partitions[pid] for pid in assignment[0][1]`.